gateway/src/apicast/policy/3scale_batcher/3scale_batcher.lua

```diff
@@ -0,0 +1,160 @@
+-- TODO:
+--
+-- - The policy might make the same call to backend several times at the same
+-- time if it receives several requests with the same credentials and those are
+-- not cached.
+--
+-- - The max number of reports that can be reported to 3scale backend in a
+-- single call is limited, but this is not respected by the current code.
+--
+-- - Timeouts in locks are not handled properly. The request simply fails when
+-- this happens.
+--
+-- - Evaluate using a local hash and semaphores in the reports batcher instead
```

Comment:
Local hash would be lost when worker crashes. Maybe using shdict but still parallelizing the collection / reporting across workers?

Using https://github.com/openresty/lua-nginx-module#ngxshareddictrpush we could immediately add reports to the list and they could be processed by several workers.

Maybe to limit congestion and improve cache locality we could have a semaphore to signal how many workers there should be running.

`rpush` returns the length of the list. We could simply call `semaphore:push(1)` when that length is over some batch size. That would wake up the `timer.every` loop waiting for the semaphore by `semaphore:wait(t)`.  When timer loop reaches empty list it can go back to `semaphore:wait()` to wait for some elements in the list.

Just food for though. There are some nice algorithms we could implement like https://elixir-lang.org/blog/2016/07/14/announcing-genstage/